### PR TITLE
Implement server lifecycle methods

### PR DESCRIPTION
**Pull request by AlpacaGoesCrazy**
This implements [server lifecycle methods available in subscriptions-transport-ws](https://github.com/apollographql/subscriptions-transport-ws#constructoroptions-socketoptions--socketserver):
- **`onConnect`** fires on `GQL_CONNECTION_INIT` message, receives payload of said message and allows to reject connection, write some data to connection context which will be available during graphql resolver execution. If `onConnect` is not defined, we put everything in `GQL_CONNECTION_INIT` message payload to connection context.
- **`onDisconnect`** fires on `$disconnect`, receives connection object as argument
- **`onOperation`** fires on mutation, query and subscription start operations, receives message payload and connection object as arguments
- **`onOperationComplete`** fires on subscription stop operations, receives connection object and operation ID as arguments

### CODE REVIEW

**Comment by AlpacaGoesCrazy** [comment]
During my research I encountered some problems which I would like to have discussed.
First of all there are two methods of authenticating websocket connections:
1. Authenticate on connect. This is achieved with the help of lambda authorizer(currently cognito authorizer is not supported) on API gateway websockets. Basicaly we provide some auth info in query string or in headers of connection request(btw browser websockets does not allow http header). If auth info is present and lambda authorizes this info, you are allowed to send messages and you can write some info to requestContext from lambda(this info will be available on subsequent calls). This method is not in the scope of the library and works poorly in browsers due to headers
2. Connect and then authorize the connection. We allow clients to connect and then we use `GQL_CONNECTION_INIT` message to get auth info and `onConnect` method to decide if we want to allow this connection. This method does have some flaws as well. In subscriptions-transport-ws server [it is not ensured that client actually sent init message](https://github.com/apollographql/subscriptions-transport-ws/issues/349), so `onConnect` might get skipped all together. And if it was sent the initialization process relies on promises to ensure that `onConnect` will be completed before the next message from client gets processed. And client application [does not wait for `GQL_CONNECTION_ACK`](https://github.com/apollographql/subscriptions-transport-ws/issues/339) to send its subscription message. So managing authorization this way in lambda is not possible because you can not preserve a Promise between lambda invocations (init and next message).

So my question is how will be better to approach the second authorization method if it is even possible in lambda environment. Would like to hear @michalkvasnicak 's input

**Comment by michalkvasnicak** [comment]
@AlpacaGoesCrazy 

Maybe we could store an internal phase to connection's data. Possible value could be `authentication | initialised` or something better?

Then if:
 - `onConnect` is **specified**
	- we set `phase = authentication` in connection store
    - all operations received in the meantime are rejected with GraphQL error until the `phase` is `initialised`
    - when `onConnect` is rejected, we terminate the connection using Connection manager
 - `onConnect` is **not specified**
	- we set `phase = initialised` in connection store
    - all operations work exactly as they are working right now because the condition about `phase = initialised` is fulfilled.

I know it's not 1:1 with Apollo's implementation in terms of operations behaviour.

Another way is to let user's to implement their own auth mechanism in the same way as in req-res model using `mutation`. So basically they will check the auth in their resolvers.

What do you think?

**Comment by AlpacaGoesCrazy** [comment]
@michalkvasnicak the problem with storing authentication | initialised in dynamoDB is that subscriptions-transport-ws does not wait for connection_ack message and sends subscription operation right away. This operation will almost always get `phase = authentication` as it is a race condition between `connection_init` message setting `phase = initialised` and subscription operation getting connection state.
Here is an example how they [deal with this in subscriptions-transport-ws](https://github.com/apollographql/subscriptions-transport-ws/blob/594ee3a6213e370ba5358858db6379ef88e6e600/src/server.ts#L306)

**Comment by michalkvasnicak** [comment]
@AlpacaGoesCrazy yes I know that there is race condition that's why I'm proposing that the server will reject all GraphQL operations until the `onConnect` is resolved.

So basically you connect to the server, `onConnect` starts and during that if you try to subscribe/query/mutate you will get GraphQLError. This is not 1:1 behaviour with how subscription-transport-ws works but it's the simplest "solution".

If you really want wait for some `initPromise`. Then `initPromise` could be just a check if connection phase is initialised (repeated fetch from DynamoDB in some interval).

**Comment by AlpacaGoesCrazy** [comment]
@michalkvasnicak I doubt that repeated fetch is a good idea in AWS Lambda environemnt

**Comment by michalkvasnicak** [comment]
@AlpacaGoesCrazy yes it isn't a good idea at all but it's the only way how to simulate the `initPromise`. Other way is just to reject operations until `initialised` phase is reached but in this case it 1:1 behaviour with subscription-transport-ws.

**Comment by AlpacaGoesCrazy** [comment]
@michalkvasnicak What is in your opinion would be a better way?

**Comment by michalkvasnicak** [comment]
@AlpacaGoesCrazy in my opinion is rejecting operations better than `initPromise`. It's easy to implement, just look to connection data if phase is initialised, if not, reject the operation with GQL error.

Another thing is that server should not implement some "batching" for client side only because the client does not wait for ACK. I think the correct way in that case should be rejecting all operations that are invoked until authentication phase is finished.

**Comment by michalkvasnicak** [comment]
@AlpacaGoesCrazy it's basically the same thing as you call a `mutation` on a resource that needs a user to be authenticated. In request-response model you'll just reject the operation with authentication error. I don't see any reason why it couldn't be that way over WebSocket too. User waits for the result anyway.

**Comment by AlpacaGoesCrazy** [comment]
@michalkvasnicak I am pretty much finished. However I still have a concern that subscriptions-transport-ws client does not wait for `connection_ack` from the server and sends subscribe operation right away. It may result in client not subscribing on the first time.

**Comment by michalkvasnicak** [comment]
@AlpacaGoesCrazy to address your concern. Yes it likely will cause first time errors if you send a subscription operation right away. But these can be addressed quite easily. However I think that your code is still fully compatible with `subscription-transport-ws` client, so when they fix the bug you mentioned it should be resolved.

Also this library is not intended only for usage with Apollo so waiting for `connection_ack` is just a protocol implementation on the client side. So for example if you use some custom client, you can easily end up with the same problem that you're sending operations during connection auth.

Does Apollo's implementation terminate the connection if authentication is rejected? Because in that case we can just terminate the connection too.

**Comment by AlpacaGoesCrazy** [comment]
@michalkvasnicak Yes, it [terminates connection](https://github.com/apollographql/subscriptions-transport-ws/blob/master/src/server.ts#L296) and my implementation does the same with `ApiGatewayManagementApi`'s `deleteConnection`

**Comment by michalkvasnicak** [comment]
Released as `0.9.0`. Thank you very much! 💯

**Comment by cookejames** [comment]
Just getting into this library (great work by the way). Perhaps you can correct me if I'm wrong but when integrating this into our existing stack I had some timing issues with the example client as subscriptions were proceeding before a connection had been made and initialised.

Adding the following in the example client fixed the problem. If this is expected behaviour let me know and I'll submit a PR as I spent a little while debugging this.

```
  useEffect(() => {
    wsClient.onConnected((...args) => {
      console.log('Connected', args)
      setConnected(true)
    })
    
    wsClient.onError((...args) => {
      console.log('Error', args)
      setConnected(false)
    })

    return () => wsClient.close()
  }, [])

  return (
    <ApolloProvider client={client}>
      {isConnected && 
         ...
```

**Comment by michalkvasnicak** [comment]
@cookejames thank you for the report, have you tried it with `0.11.0` version or `0.10.0`? 

It's caused basically by the design of Apollo's protocol. They don't wait for ACK and fire subscriptions immediately. In version `0.11.0` there are options to wait for connection to be initialised for some time until the `Prohibited connection` is fired. 

```ts
const wsHandler = createWsHandler({
  connectionManager,
  schema,
  subscriptionManager,
  // validationRules
  // waitForInitialization (available from 0.11.0)
  // waitForInitialization: { retryCount?: number, timeout?: number },
});
```

or see this file https://github.com/michalkvasnicak/aws-lambda-graphql/blob/master/packages/aws-lambda-graphql/src/createWsHandler.ts#L58 and the defaults are here https://github.com/michalkvasnicak/aws-lambda-graphql/blob/master/packages/aws-lambda-graphql/src/createWsHandler.ts#L86.

Basically when connection is not initialised, it'll try to hydrate the connection state after `waitForInitialization.timeout` and will repeat that for `waitForInitialization.retryCount` until `Prohibited connection` is thrown.

**Comment by cookejames** [comment]
@michalkvasnicak thanks for the pointer I was on 0.10.0 and that change worked without the need to check for the connection event
